### PR TITLE
80835  Export IPOs to excel

### PR DESCRIPTION
**Pull request by bratli**
[AB#80835](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/80835)

Export IPOs to excel endpoint + tests

If we export only one IPO, then we get participant list and history for IPO as well.

### CODE REVIEW

**Comment by janid1967** [comment]
I'll review today ... have you seen red build @bratli ?

File: src/Equinor.ProCoSys.IPO.Query/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExcelQueryHandler.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            {
+                return new SuccessResult<ExportDto>(new ExportDto(null, usedFilterDto));
+            }
+
+            var invitationIds = orderedDtos.Select(dto => dto.Id).ToList();
+            var getHistoryAndParticipants = invitationIds.Count == 1;
+
+            var invitationsWithIncludes =
+                await GetInvitationsWithIncludesAsync(invitationIds, getHistoryAndParticipants, cancellationToken);
+
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExcelQueryHandler.cs:49]: remove

File: src/Equinor.ProCoSys.IPO.Query/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExcelQueryHandler.cs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        private async Task<string> GetPersonNameAsync(Guid? personOid)
+        {
+            if (!personOid.HasValue)
+            {
+                return null;
+            }
+
+            return await (from p in _context.QuerySet<Person>()
+                where p.Oid == personOid
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExcelQueryHandler.cs:157]: personOid.Value

File: src/Equinor.ProCoSys.IPO.Query/GetInvitationsQueries/Sorting.cs [hunk 3]
```diff
@@ -1,4 +1,6 @@
-﻿namespace Equinor.ProCoSys.IPO.Query.GetInvitations
+﻿using Equinor.ProCoSys.IPO.Query.GetInvitations;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetInvitationsQueries/Sorting.cs:1]: remove

**Comment by janid1967** [comment]
PunchOutDateFilterType.cs has wrong ns

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/PagingDto.cs [hunk 4]
```diff
@@ -1,4 +1,5 @@
 ﻿using Equinor.ProCoSys.IPO.Query.GetInvitations;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/PagingDto.cs:1]: remove

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/SortingDto.cs [hunk 5]
```diff
@@ -1,4 +1,6 @@
 ﻿using Equinor.ProCoSys.IPO.Query.GetInvitations;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/SortingDto.cs:1]: remove

File: src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+using Equinor.ProCoSys.IPO.Query.GetInvitationsQueries.GetInvitationsForExport;
+
+namespace Equinor.ProCoSys.IPO.WebApi.Excel
+{
+    public class ExcelConverter : IExcelConverter
+    {
+        public static class FrontSheetRows
+        {
+            public static int MainHeading = 1;
+            public static int Blank1 = 2;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs:15]: remove Blank1 and Blank2 and reorder

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs [hunk 7]
```diff
@@ -0,0 +1,767 @@
...
+
+        [TestMethod]
+        public async Task HandleGetInvitationsForExportQuery_ShouldSortOnContractorRep_Asc()
+        {
+            var sorting = new Sorting(SortingDirection.Asc, SortingProperty.ContractorRep);
+
+            using (var context =
+                new IPOContext(_dbContextOptions, _plantProvider, _eventDispatcher, _currentUserProvider))
+            {
+                var query = new GetInvitationsForExportQuery(_projectName2, sorting, null);
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs:626]: remove last param (null is default)

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs [hunk 8]
```diff
@@ -0,0 +1,767 @@
...
+
+        [TestMethod]
+        public async Task HandleGetInvitationsForExportQuery_ShouldSortOnType_Asc()
+        {
+            var sorting = new Sorting(SortingDirection.Asc, SortingProperty.Type);
+
+            using (var context =
+                new IPOContext(_dbContextOptions, _plantProvider, _eventDispatcher, _currentUserProvider))
+            {
+                var query = new GetInvitationsForExportQuery(_projectName2, sorting, null);
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs:662]: remove last param (null is default)

File: src/tests/Equinor.ProCoSys.IPO.Test.Common/ReadOnlyTestsBase.cs [hunk 9]
```diff
@@ -1,11 +1,15 @@
...
 using System.Linq;
 using Equinor.ProCoSys.IPO.Domain;
+using Equinor.ProCoSys.IPO.Domain.AggregateModels.HistoryAggregate;
+using Equinor.ProCoSys.IPO.Domain.AggregateModels.InvitationAggregate;
 using Equinor.ProCoSys.IPO.Domain.AggregateModels.PersonAggregate;
 using Equinor.ProCoSys.IPO.Domain.Events;
 using Equinor.ProCoSys.IPO.Domain.Time;
 using Equinor.ProCoSys.IPO.Infrastructure;
 using Microsoft.EntityFrameworkCore;
+using Microsoft.VisualBasic.CompilerServices;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Test.Common/ReadOnlyTestsBase.cs:12]: wow

File: src/tests/Equinor.ProCoSys.IPO.Test.Common/ReadOnlyTestsBase.cs [hunk 10]
```diff
@@ -1,11 +1,15 @@
 ﻿using System;
+using System.Collections.Generic;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Test.Common/ReadOnlyTestsBase.cs:2]: remove unused usings

File: src/tests/Equinor.ProCoSys.IPO.WebApi.IntegrationTests/Invitations/InvitationsControllerTestsHelper.cs [hunk 11]
```diff
@@ -3,9 +3,12 @@
 using System.Net.Http;
 using System.Text;
 using System.Threading.Tasks;
+using ClosedXML.Excel;
 using Equinor.ProCoSys.IPO.Command.InvitationCommands;
 using Equinor.ProCoSys.IPO.Query.GetInvitationsByCommPkgNo;
 using Equinor.ProCoSys.IPO.Query.GetLatestMdpIpoStatusOnCommPkgs;
+using Equinor.ProCoSys.IPO.WebApi.Controllers.Invitation;
+using Microsoft.AspNetCore.Mvc;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.WebApi.IntegrationTests/Invitations/InvitationsControllerTestsHelper.cs:11]: remove

**Review by janid1967 — CHANGES_REQUESTED** [review]
See comments

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/PagingDto.cs (hunk 4 above)
Comment (bratli) [reply src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/PagingDto.cs:1]: remove the file ?

Comment (janid1967) [reply src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/PagingDto.cs:1]: wasn't it a unused using?

Comment (bratli) [reply src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/PagingDto.cs:1]: Don't see any unused usings now at least

File: src/tests/Equinor.ProCoSys.IPO.Test.Common/ReadOnlyTestsBase.cs (hunk 9 above)
Comment (bratli) [reply src/tests/Equinor.ProCoSys.IPO.Test.Common/ReadOnlyTestsBase.cs]: :(

**Review by janid1967 — APPROVED** [review]

**Pull request merged**